pgrx/src/list.rs

```diff
@@ -12,257 +12,617 @@
 //! It functions similarly to a Rust [`Vec`][std::vec::Vec], including iterator support, but provides separate
 //! understandings of [`List`][crate::pg_sys::List]s of [`Oid`][crate::pg_sys::Oid]s, Integers, and Pointers.
 
-use crate::{is_a, pg_sys, void_mut_ptr};
-use std::marker::PhantomData;
+#[cfg(any(feature = "pg13", feature = "pg14", feature = "pg15", feature = "pg16"))]
+pub use flat_list::{Enlist, List, ListHead};
+
+#[cfg(any(feature = "pg13", feature = "pg14", feature = "pg15", feature = "pg16"))]
+mod flat_list {
+    use crate::pg_sys;
+    use core::cmp;
+    use core::ffi;
+    use core::fmt::{self, Debug, Formatter};
+    use core::marker::PhantomData;
+    use core::mem;
+    use core::ops::{Bound, Deref, DerefMut, RangeBounds};
+    use core::ptr::{self, NonNull};
+    use core::slice;
+
+    /// The List type from Postgres, lifted into Rust
+    /// Note: you may want the ListHead type
+    #[derive(Debug)]
+    pub enum List<T> {
+        Nil,
+        Cons(ListHead<T>),
+    }
 
-pub struct PgList<T> {
-    list: *mut pg_sys::List,
-    allocated_by_pg: bool,
-    _marker: PhantomData<T>,
-}
-impl<T> Default for PgList<T> {
-    fn default() -> Self {
-        Self::new()
+    /// A strongly-typed ListCell
+    #[repr(transparent)]
+    pub struct ListCell<T> {
+        // It is important that we are able to treat this union as effectively synonymous with T!
+        // Thus it is important that we
+        // - do not hand out the ability to construct arbitrary ListCell<T>
+        // - do not offer casting between types of List<T> (which offer [ListCell<T>])
+        // - do not even upgrade from pg_sys::{List, ListCell} to pgrx::list::{List, ListCell}
+        // UNLESS the relevant safety invariants are appropriately handled!
+        // It is not even okay to do this for FFI! We must check any *mut pg_sys::List from FFI,
+        // to guarantee it has the expected type tag, otherwise the union cells may be garbage.
+        cell: pg_sys::ListCell,
+        _type: PhantomData<T>,
     }
-}
 
-impl<T> PgList<T> {
-    pub fn new() -> Self {
-        PgList {
-            list: std::ptr::null_mut(), // an empty List is NIL
-            allocated_by_pg: false,
-            _marker: PhantomData,
+    impl<T> Debug for ListCell<T> {
+        fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+            // interpret the cell as a pointer, assuming pointers are Pod
+            let cell_addr = unsafe { self.cell.ptr_value };
+            f.debug_tuple("ListCell").field(&cell_addr).finish()
         }
     }
 
-    pub unsafe fn from_pg(list: *mut pg_sys::List) -> Self {
-        PgList { list, allocated_by_pg: true, _marker: PhantomData }
+    // Note this well: the size of a `ListCell<T>`'s type doesn't matter.
+    // Thus it isn't acceptable to implement Enlist for types larger than `pg_sys::ListCell`.
```

Comment:
```suggestion
```suggestion
    // Note: the size of `ListCell<T>`'s generic `T` doesn't matter,
    // thus it isn't acceptable to implement Enlist for a `T` larger than `pg_sys::ListCell`.
```